drivers/crypto/crypto_mtls_shim.c

```diff
@@ -171,13 +222,364 @@ static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
+	mbedtls_operation_t op;
+	int r;
+
+	if (!(mode & CRYPTO_CIPHER_MODE_CMAC)) {
+		SYS_LOG_ERR("Only CMAC supported for MAC algo");
+		return -EINVAL;
+	}
+
+	if (mode & CRYPTO_CIPHER_MODE_ECB) {
+		type = MBEDTLS_CIPHER_AES_128_ECB;
+	} else if (mode & CRYPTO_CIPHER_MODE_CBC) {
+		type = MBEDTLS_CIPHER_AES_128_CBC;
+	} else if (mode & CRYPTO_CIPHER_MODE_CTR) {
+		type = MBEDTLS_CIPHER_AES_128_CTR;
+	} else if (mode & CRYPTO_CIPHER_MODE_CCM) {
+		type = MBEDTLS_CIPHER_AES_128_CCM;
+	} else {
+		/* Defaults to AES, 128-bit keys, ECB mode. */
+		type = MBEDTLS_CIPHER_AES_128_ECB;
+	}
+	ci = mbedtls_cipher_info_from_type(type);
+	if (!ci) {
+		return -EINVAL;
+	}
+
+	session = mtls_alloc_session(CRYPTO_CIPHER_ALGO_MAC);
+	if (!session) {
+		return -ENOSPC;
+	}
+
+	mbedtls_cipher_init(&session->cipher);
+
+	r = mbedtls_cipher_setup(&session->cipher, ci);
+	if (r != 0) {
+		mtls_free_session(session);
+		return -EINVAL;
+	}
+
+	if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
+		op = MBEDTLS_ENCRYPT;
+	} else {
+		op = MBEDTLS_DECRYPT;
+	}
+
+	r = mbedtls_cipher_setkey(&session->cipher,
+				  ctx->key.bit_stream, 128, op);
+	if (r != 0) {
+		mtls_free_session(session);
+		return -EINVAL;
+	}
+
+	ctx->ops.mac_crypt_hndlr = mtls_mac_op;
+	ctx->ops.cipher_mode = mode;
+	ctx->drv_sessn_state = session;
+
+	return 0;
+}
+
+static int mtls_ecc_gen_key(struct cipher_ctx *ctx,
+			    struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_keypair keypair;
+	size_t olen;
+	int r;
+
+	ARG_UNUSED(ctx);
+
+	if (pkt->curve != CIPHER_ECC_CURVE_SECP256R1) {
+		return -EINVAL;
+	}
+
+	mbedtls_ecp_keypair_init(&keypair);
+	r = mbedtls_ecp_gen_key(MBEDTLS_ECP_DP_SECP256R1, &keypair,
+				mtls_get_entropy_from_device, NULL);
+	if (r != 0) {
+		goto free_keypair;
+	}
+
+	r = mbedtls_mpi_write_binary(&keypair.d,
+				     pkt->private_key,
+				     pkt->private_key_len);
+	if (r != 0) {
+		goto free_keypair;
+	}
+
+	r = mbedtls_ecp_point_write_binary(&keypair.grp,
+					   &keypair.Q,
+					   MBEDTLS_ECP_PF_UNCOMPRESSED,
+					   &olen,
+					   pkt->private_key,
+					   pkt->private_key_len);
+free_keypair:
+	mbedtls_ecp_keypair_free(&keypair);
+
+	return r == 0 ? 0 : -EIO;
+}
+
+static int mtls_ecc_validate_pubkey(struct cipher_ctx *ctx,
+				    struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_group grp;
+	mbedtls_ecp_point pt;
+	int r;
+
+	ARG_UNUSED(ctx);
+
+	if (pkt->curve != CIPHER_ECC_CURVE_SECP256R1) {
+		return -EINVAL;
+	}
+
+	r = mbedtls_ecp_group_load(&grp, MBEDTLS_ECP_DP_SECP256R1);
+	if (r != 0) {
+		return -EINVAL;
+	}
+
+	mbedtls_ecp_point_init(&pt);
+	r = mbedtls_ecp_point_read_binary(&grp, &pt,
+					  pkt->public_key,
+					  pkt->public_key_len);
+	if (r != 0) {
+		r = mbedtls_ecp_check_pubkey(&grp, &pt);
+	}
+
+	mbedtls_ecp_point_free(&pt);
+	mbedtls_ecp_group_free(&grp);
+
+	return r == 0 ? 0 : -EINVAL;
+}
+
+static int mtls_ecc_gen_shared_key(struct cipher_ctx *ctx,
+				   struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_keypair keypair;
+	mbedtls_ecp_point Q;
```

Comment:
question, why this variable is uppercase ?

Reply:
This is what mbedtls uses; I used this for consistency.  That's also the same notation used in ECDH math.  I guess I could change this to `public_key`, but then it would be inconsistent with `struct mbedtls_ecp_keypair` and other related types

Reply:
:+1: